packages/generator-langium/templates/test/test/linking/linking.test.ts

```diff
@@ -0,0 +1,53 @@
+import { afterEach, beforeAll, describe, expect, test } from "vitest";
+import { EmptyFileSystem, type LangiumDocument } from "langium";
+import { expandToString as s} from "langium/generate";
+import { clearDocuments, parseHelper } from "langium/test";
+import { create<%= LanguageName %>Services } from "../../src/language/<%= language-id %>-module.js";
+import { Model, isModel } from "../../src/language/generated/ast.js";
+
+let services: ReturnType<typeof create<%= LanguageName %>Services>;
+let parse:    ReturnType<typeof parseHelper<Model>>;
+let document: LangiumDocument<Model> | undefined;
+
+beforeAll(async () => {
+    services = create<%= LanguageName %>Services(EmptyFileSystem);
+    parse = parseHelper<Model>(services.<%= LanguageName %>);
+
+    // activate the following if your linking test requires elements from a built-in library, for example
+    // await services.shared.workspace.WorkspaceManager.initializeWorkspace([]);
+});
+
+afterEach(async () => {
+    document && clearDocuments(services, [ document ]);
+});
+
+describe('Linking tests', () => {
+
+    test('linking of greetings', async () => {
+        document = await parse(`
+            person Langium
+            Hello Langium!
+        `);
+
+        expect(
+            // here we first check for validity of the parsed document object by means of the reusable function
+            //  'checkDocumentValid()' to sort out (critical) typos first,
+            // and then evaluate the cross references we're interested in by checking
+            //  the referenced AST element as well as for a potential error message;
+            checkDocumentValid(document)
+                || document.parseResult.value.greetings.map(g => g.person.ref?.name || g.person.error?.message).join('\n')
+        ).toBe(s`
+            Langium
+        `);
+    });
+});
+
+function checkDocumentValid(document: LangiumDocument): string | undefined {
```

Comment:
This could be a function in 'langium/test'...

I am normally writing it like this:

```ts
function expectNoErrors(doc: LangiumDocument) {
  expect(doc.parseResult.lexerErrors, firstOrUndefined(doc.parseResult.lexerErrors, e => e.message)).toHaveLength(0);
  expect(doc.parseResult.parserErrors, firstOrUndefined(doc.parseResult.parserErrors, e => e.message)).toHaveLength(0);
  expect(doc.diagnostics ?? [], firstOrUndefined(doc.diagnostics ?? [], e => e.message)).toHaveLength(0);
}
function firstOrUndefined<T>(array: T[]|undefined, convert: (t: T) => string): string|undefined {
  return (array?.length??0) > 0 ? convert(array![0]) : undefined;
}
```

You do not have to use it, my hint is: do not so much in one command and split into multiple expects if possible to increase readability.

Reply:
> split into multiple expects if possible to increase readability

Maybe it simplifies reviewing the test, ok.
But does it improve anything when you run into spontaneous test failures (while working at something totally different)?

Reply:
And yes, I want to offer a convenient solution to developers that cannot rely on `vitest`'s flavor of `expect`.
Maybe I'm to kind at this point...

Reply:
I like the idea of having a generic checker for document validity. I also like the cool error formatting in this case. It is also good to not bind the user to the testing framework.
When you want to offer it to the user, it should live somewhere in our `langium/test` namespace (can also be a separate PR if you want). Otherwise we invent this wheel again and again.
Plus, I was wondering, why only parser errors? We also have lexer and validation errors.